# Investigate Hotel Business using Data Visualization (Rakamin Mini Project)

## Data Preprocessing

### Import Data and Library

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib import rcParams
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv('hotel_bookings_data.csv')
pd.set_option('display.max_columns', None)
df.sample(5)

hotel  is_canceled  lead_time  arrival_date_year  \
108008    City Hotel            0         97               2019   
9540    Resort Hotel            1         22               2018   
76023     City Hotel            1        336               2017   
20451   Resort Hotel            0        303               2018   
48598     City Hotel            1         78               2018   

       arrival_date_month  arrival_date_week_number  \
108008                May                        11   
9540             February                        50   
76023            November                        38   
20451           September                        29   
48598                 May                        13   

        arrival_date_day_of_month  stays_in_weekend_nights  \
108008                         16                        2   
9540                            8                        0   
76023                          18                        0   
20451                          13                        0   
48598                          25                        1   

        stays_in_weekdays_nights  adults  children  babies       meal  \
108008                         3       2       0.0       0  Breakfast   
9540                           3       2       0.0       0  Undefined   
76023                          2       2       0.0       0     Dinner   
20451                          4       2       2.0       0  Breakfast   
48598                          2       2       0.0       0  Breakfast   

                     city market_segment distribution_channel  \
108008  Kabupaten Bandung  Offline TA/TO                TA/TO   
9540        Kota Denpasar         Groups                TA/TO   
76023       Kota Denpasar  Offline TA/TO                TA/TO   
20451       Kota Denpasar      Online TA                TA/TO   
48598       Kota Denpasar  Offline TA/TO                TA/TO   

        is_repeated_guest  previous_cancellations  \
108008                  0                       0   
9540                    0                       0   
76023                   0                       1   
20451                   0                       0   
48598                   0                       0   

        previous_bookings_not_canceled  booking_changes deposit_type  agent  \
108008                               0                1   No Deposit   27.0   
9540                                 0                1   No Deposit   38.0   
76023                                0                0   Non Refund    6.0   
20451                                1                0   No Deposit  240.0   
48598                                0                0   Non Refund  154.0   

        company  days_in_waiting_list customer_type     adr  \
108008      NaN                     0      Personal   68.00   
9540        NaN                     0        Family   80.00   
76023       NaN                     0        Family  101.50   
20451       NaN                     0      Personal  159.38   
48598       NaN                     0      Personal  100.00   

        required_car_parking_spaces  total_of_special_requests  \
108008                            0                          0   
9540                              0                          0   
76023                             0                          0   
20451                             1                          2   
48598                             0                          0   

       reservation_status  
108008          Check-Out  
9540             Canceled  
76023            Canceled  
20451           Check-Out  
48598            Canceled

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_weekdays_nights        119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

### Missing Value Treatment

In [4]:
df.isna().sum().sort_values(ascending=False)

company                           112593
agent                              16340
city                                 488
children                               4
hotel                                  0
is_repeated_guest                      0
total_of_special_requests              0
required_car_parking_spaces            0
adr                                    0
customer_type                          0
days_in_waiting_list                   0
deposit_type                           0
booking_changes                        0
previous_bookings_not_canceled         0
previous_cancellations                 0
market_segment                         0
distribution_channel                   0
is_canceled                            0
meal                                   0
babies                                 0
adults                                 0
stays_in_weekdays_nights               0
stays_in_weekend_nights                0
arrival_date_day_of_month              0
arrival_date_wee

In [5]:
sum_missing = df.isna().sum()
percent_missing = round(df.isnull().sum() * 100 / len(df),4)
missing_value_df = pd.DataFrame({'column_name': df.columns,
                                 'sum_missing': sum_missing,
                                 'percent_missing' : percent_missing}).reset_index(drop=True).sort_values(by=['percent_missing','sum_missing'], ascending=False)
missing_value_df.head(4)

column_name  sum_missing  percent_missing
22     company       112593          94.3069
21       agent        16340          13.6862
13        city          488           0.4087
10    children            4           0.0034

#### What the best treatment for the missing values?
- The `company` column contains null values of as many as **112593 (94.31%)** . One thing is possible I can drop the column however, there is a chance I will lose the **vital information of 5.69%** of the data in that column. So the best treatment is to **replace the null values with 0**. This also applies to the `agent` column.
- For `city` with 488 missing values(0.41%). I'll replace in the missing values with `'Not Avaliable'`
- Last but not least, I will drop 4 missing values in the `children` column.

In [6]:
df2 = df.copy()

In [7]:
val = {'company': 0, 'agent': 0, 'city':'Not Avaliable'}
df2.fillna(value = val, inplace = True)

df2.dropna(subset = ['children'], inplace = True)

In [8]:
df2.isna().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_weekdays_nights          0
adults                            0
children                          0
babies                            0
meal                              0
city                              0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests   

### Incosistent and Invalid Data

In [9]:
df3 = df2.copy()

In [10]:
#Grouping the categorical features
cats = ['hotel', 'is_canceled', 'meal', 'city',
        'market_segment', 'distribution_channel',
        'is_repeated_guest', 'deposit_type',
        'customer_type', 'reservation_status']

In [11]:
for col in cats:
    unique_values = df3[col].unique()
    print(f'''Value count column {col}:''')
    print(f"{unique_values}")
    print()

Value count column hotel:
['Resort Hotel' 'City Hotel']

Value count column is_canceled:
[0 1]

Value count column meal:
['Breakfast' 'Full Board' 'Dinner' 'No Meal' 'Undefined']

Value count column city:
['Kota Denpasar' 'Kabupaten Bangka' 'Kabupaten Sleman' 'Kota Batu'
 'Kota Malang' 'Kota Yogyakarta' 'Not Avaliable' 'Kabupaten Tangerang'
 'Kota Semarang' 'Kota Serang' 'Kota Tangerang' 'Kota Bogor'
 'Kabupaten Bandung' 'Kabupaten Magelang' 'Kabupaten Sumedang'
 'Kota Jakarta Timur' 'Kabupaten Purwakarta' 'Kabupaten Kepulauan Seribu'
 'Kabupaten Belitung' 'Kota Jakarta Pusat' 'Kota Jakarta Selatan'
 'Kota Surabaya' 'Kota Jakarta Utara' 'Kota Gorontalo' 'Kota Jambi'
 'Kota Jakarta Barat' 'Kota Bengkulu' 'Kabupaten Bandung Barat'
 'Kabupaten Bekasi' 'Kabupaten Bogor' 'Kabupaten Ciamis'
 'Kabupaten Cianjur' 'Kabupaten Cirebon' 'Kota Cimahi'
 'Kabupaten Indramayu' 'Kabupaten Karawang' 'Kabupaten Kuningan'
 'Kabupaten Majalengka' 'Kabupaten Serang' 'Kota Pangkal Pinang'
 'Kabupaten Subang'

In [12]:
df3['meal'].value_counts().to_frame()

meal
Breakfast   92306
Dinner      14463
No Meal     10650
Undefined    1169
Full Board    798

In [13]:
df3['distribution_channel'].value_counts().to_frame()

distribution_channel
TA/TO                     97870
Direct                    14645
Corporate                  6677
GDS                         193
Undefined                     1

#### Inconsistence Value
There are 2 columns that have the value `'Undefined'`, namely the `meal` column and `distribution_channel`. For the `meal` column, I will replace it with `'No Meal'` because it more or less has the same meaning.
</br> Whereas in the `distribution_channel` column, because there is only one row with `'Undefined'` then I will drop the row

In [14]:
df3.drop(df3[df3['distribution_channel'] == 'Undefined'].index, inplace=True)
df3['meal'].replace('Undefined', 'No Meal', inplace=True)

In [15]:
df3['meal'].unique()

array(['Breakfast', 'Full Board', 'Dinner', 'No Meal'], dtype=object)

In [16]:
df3['distribution_channel'].unique()

array(['Direct', 'Corporate', 'TA/TO', 'GDS'], dtype=object)

#### Invalid Value

In [17]:
df4 = df3.copy()

In [18]:
df4.describe()

is_canceled      lead_time  arrival_date_year  \
count  119385.000000  119385.000000      119385.000000   
mean        0.370398     104.014809        2018.156603   
std         0.482913     106.863734           0.707451   
min         0.000000       0.000000        2017.000000   
25%         0.000000      18.000000        2018.000000   
50%         0.000000      69.000000        2018.000000   
75%         1.000000     160.000000        2019.000000   
max         1.000000     737.000000        2019.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119385.000000              119385.000000   
mean                  27.164996                  15.798643   
std                   13.605391                   8.780765   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_weekdays_nights         adults  \
count            119385.000000             119385.000000  119385.000000   
mean                  0.927596                  2.500306       1.856389   
std                   0.998618                  1.908296       0.579263   
min                   0.000000                  0.000000       0.000000   
25%                   0.000000                  1.000000       2.000000   
50%                   1.000000                  2.000000       2.000000   
75%                   2.000000                  3.000000       2.000000   
max                  19.000000                 50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119385.000000  119385.000000      119385.000000   
mean        0.103882       0.007949           0.031914   
std         0.398555       0.097438           0.175771   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119385.000000                   119385.000000   
mean                 0.087121                        0.137103   
std                  0.844354                        1.497468   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent        company  days_in_waiting_list  \
count    119385.000000  119385.000000  119385.000000         119385.000000   
mean          0.221133      74.831260      10.775608              2.321246   
std           0.652318     107.143226      53.944968             17.595083   
min           0.000000       0.000000       0.000000              0.000000   
25%           0.000000       7.000000       0.000000              0.000000   
50%           0.000000       9.000000       0.000000              0.000000   
75%           0.000000     152.000000       0.000000              0.000000   
max          21.000000     535.000000     543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119385.000000                119385.000000              119385.000000  
mean      101.833450                     0.062512                   0.571328  
std        50.534866                     0.245281                   0.792791  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.0

What I wat to underline here is the `adults` column where the MIN value is 0, which is unlikely because children cannot book rooms without adults. So, I'm going to drop rows where the `adults` column is 0.

### Feature Engineering

Two columns can be added to the existing dataset, namely:
1. `Total_guests` by `adults` + `kids` (`babies` + `children`)
2. `stay_duration` by `stays_in_weekend_nights` + `stays_in_weekdays_nights`

#### Total_guests

In [19]:
df6 = df4.copy()

In [20]:
#Changing datatype from float to int
df6['children'] = df6['children'].astype(int)

# Create kids column
df6['kids']=df6['children']+df6['babies']

df6['total_guests'] = df6['adults'] + df6['kids']

# dropping rows that include 0 Total Guests
df6 = df6[df6['total_guests'] != 0]

#### Stay_duration

In [21]:
df7 = df6.copy()

In [22]:
# Creating 'total_nights'
df7['total_nights'] = df7['stays_in_weekend_nights'] + df7['stays_in_weekdays_nights']

In [23]:
df7.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 119205 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119205 non-null  object 
 1   is_canceled                     119205 non-null  int64  
 2   lead_time                       119205 non-null  int64  
 3   arrival_date_year               119205 non-null  int64  
 4   arrival_date_month              119205 non-null  object 
 5   arrival_date_week_number        119205 non-null  int64  
 6   arrival_date_day_of_month       119205 non-null  int64  
 7   stays_in_weekend_nights         119205 non-null  int64  
 8   stays_in_weekdays_nights        119205 non-null  int64  
 9   adults                          119205 non-null  int64  
 10  children                        119205 non-null  int32  
 11  babies                          119205 non-null  int64  
 12  meal            

## Analysis 1

## Monthly Hotel Booking Analysis Based on Hotel Type

In [24]:
df_resort = df7[(df7['hotel'] == 'Resort Hotel') & (df7['is_canceled'] == 0)]
df_city = df7[(df7['hotel'] == 'City Hotel') & (df7['is_canceled'] == 0)]

In [25]:
resort_total = df_resort.groupby(['arrival_date_month'])['total_guests'].count().reset_index()
resort_total

arrival_date_month  total_guests
0               April          2308
1              August          2037
2            December          2575
3            February          2014
4             January          1975
5                July          2535
6                June          2550
7               March          1866
8                 May          2571
9            November          2102
10            October          3257
11          September          3136

In [26]:
city_total = df_city.groupby(['arrival_date_month'])['total_guests'].count().reset_index()
city_total

arrival_date_month  total_guests
0               April          3051
1              August          4358
2            December          4326
3            February          2377
4             January          2676
5                July          4568
6                June          4010
7               March          2249
8                 May          4049
9            November          4283
10            October          5367
11          September          4770

In [27]:
total_hotel = resort_total.merge(city_total, on = 'arrival_date_month')
total_hotel.columns = ['month', 'total_bookings_resort', 'total_bookings_city_hotel']
total_hotel

month  total_bookings_resort  total_bookings_city_hotel
0       April                   2308                       3051
1      August                   2037                       4358
2    December                   2575                       4326
3    February                   2014                       2377
4     January                   1975                       2676
5        July                   2535                       4568
6        June                   2550                       4010
7       March                   1866                       2249
8         May                   2571                       4049
9    November                   2102                       4283
10    October                   3257                       5367
11  September                   3136                       4770

In [28]:
import sort_dataframeby_monthorweek as sd

def sort_month(df, column_name):
    return sd.Sort_Dataframeby_Month(df, column_name)

In [29]:
total_bookings = sort_month(total_hotel, 'month')
total_bookings

month  total_bookings_resort  total_bookings_city_hotel
0     January                   1975                       2676
1    February                   2014                       2377
2       March                   1866                       2249
3       April                   2308                       3051
4         May                   2571                       4049
5        June                   2550                       4010
6        July                   2535                       4568
7      August                   2037                       4358
8   September                   3136                       4770
9     October                   3257                       5367
10   November                   2102                       4283
11   December                   2575                       4326

In [30]:
booking_sep = total_bookings[(total_bookings['month'] == 'September')]
booking_sep

month  total_bookings_resort  total_bookings_city_hotel
8  September                   3136                       4770

In [31]:
booking_oct = total_bookings[(total_bookings['month'] == 'October')]
booking_oct

month  total_bookings_resort  total_bookings_city_hotel
9  October                   3257                       5367

In [32]:
booking_sep = booking_sep.append(booking_oct, ignore_index=True)
booking_sep

month  total_bookings_resort  total_bookings_city_hotel
0  September                   3136                       4770
1    October                   3257                       5367

In [33]:
booking_sep.drop_duplicates(inplace = True)
booking_sep

month  total_bookings_resort  total_bookings_city_hotel
0  September                   3136                       4770
1    October                   3257                       5367

In [34]:
booking_3m = booking_sep.copy()
booking_3m['avg_booking_resort'] = round((booking_sep['total_bookings_resort']/3),0)
booking_3m['avg_booking_city_hotel'] = round((booking_sep['total_bookings_city_hotel']/3),0)
booking_3m

month  total_bookings_resort  total_bookings_city_hotel  \
0  September                   3136                       4770   
1    October                   3257                       5367   

   avg_booking_resort  avg_booking_city_hotel  
0              1045.0                  1590.0  
1              1086.0                  1789.0

In [35]:
del booking_3m['total_bookings_resort']
del booking_3m['total_bookings_city_hotel']

In [36]:
avg_booking = total_bookings.copy()

In [37]:
avg_booking.drop(avg_booking[avg_booking['month'] == 'September'].index, inplace=True)
avg_booking.drop(avg_booking[avg_booking['month'] == 'October'].index, inplace=True)

In [38]:
avg_booking['avg_booking_resort'] = round((avg_booking['total_bookings_resort']/2),0)
avg_booking['avg_booking_city_hotel'] = round((avg_booking['total_bookings_city_hotel']/2),0)

In [39]:
del avg_booking['total_bookings_resort']
del avg_booking['total_bookings_city_hotel']
avg_booking

month  avg_booking_resort  avg_booking_city_hotel
0    January               988.0                  1338.0
1   February              1007.0                  1188.0
2      March               933.0                  1124.0
3      April              1154.0                  1526.0
4        May              1286.0                  2024.0
5       June              1275.0                  2005.0
6       July              1268.0                  2284.0
7     August              1018.0                  2179.0
10  November              1051.0                  2142.0
11  December              1288.0                  2163.0

In [40]:
avg_booking = avg_booking.append(booking_3m, ignore_index=True)
avg_booking

month  avg_booking_resort  avg_booking_city_hotel
0     January               988.0                  1338.0
1    February              1007.0                  1188.0
2       March               933.0                  1124.0
3       April              1154.0                  1526.0
4         May              1286.0                  2024.0
5        June              1275.0                  2005.0
6        July              1268.0                  2284.0
7      August              1018.0                  2179.0
8    November              1051.0                  2142.0
9    December              1288.0                  2163.0
10  September              1045.0                  1590.0
11    October              1086.0                  1789.0

In [41]:
avg_bookings = sort_month(avg_booking, 'month')
avg_bookings['avg_booking_resort'] = avg_bookings['avg_booking_resort'].astype(int)
avg_bookings['avg_booking_city_hotel'] = avg_bookings['avg_booking_city_hotel'].astype(int)

In [42]:
import plotly.express as px

In [43]:
plt.figure(figsize = (17, 8))

px.line(avg_bookings, x = 'month', y = ['avg_booking_resort','avg_booking_city_hotel'],
        title = 'Average Monthly Hotel Booking Analysis Based on Hotel Type', template = 'plotly_dark')

<Figure size 1700x800 with 0 Axes>